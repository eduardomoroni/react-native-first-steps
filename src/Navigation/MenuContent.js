// @flow

import React, { Component } from 'react'
import { ScrollView, Image, BackAndroid } from 'react-native'
import { connect } from 'react-redux'
import styles from './Styles/MenuContentStyle'
import { Images } from '../Themes'
import { userLoggout } from '../Redux/Actions'
import DrawerButton from '../Containers/Components/MenuButton'
import { Actions as NavigationActions } from 'react-native-router-flux'
import I18n from 'react-native-i18n'

class DrawerContent extends Component {
  componentDidMount () {
    BackAndroid.addEventListener('hardwareBackPress', () => {
      if (this.context.drawer.props.open) {
        this.toggleDrawer()
        return true
      }
      return false
    })
  }

  toggleDrawer () {
    this.context.drawer.toggle()
  }

  handlePress = (route: string) => {
    this.toggleDrawer()
    switch (route) {
      case 'settings':
        NavigationActions.settings()
        break
      case 'cardSearch':
        NavigationActions.cardSearch()
        break
      case 'loginScreen':
        NavigationActions.loginScreen()
        break
    }
  }

  renderLoginOrLogout = () => {
    if (this.props.user === null) {
      return (
        <DrawerButton text={I18n.t('login')} onPress={() => this.handlePress('loginScreen')} />
      )
    } else {
      return (
        <DrawerButton text={I18n.t('logout')} onPress={() => this.props.onLogout()} />
      )
    }
  }

  render () {
    return (
      <ScrollView style={styles.container}>
        <Image source={Images.logo} style={styles.logo} />
        <DrawerButton text={I18n.t('settings')} onPress={() => this.handlePress('settings')} />
        {this.renderLoginOrLogout()}
      </ScrollView>
    )
  }
}

DrawerContent.contextTypes = {
  drawer: React.PropTypes.object
}

const mapStateToProps = (state) => {
  const { user } = state.user
  return {
    user: user
  }
}

const mapDispatchToProps = (dispatch) => {
  return {
    onLogout: () => dispatch(userLoggout())
  }
}

export default connect(mapStateToProps, mapDispatchToProps)(DrawerContent)
