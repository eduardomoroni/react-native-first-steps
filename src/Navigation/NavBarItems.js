/* @flow */

import React from 'react'
import {
  StyleSheet,
  TouchableOpacity
} from 'react-native'
import { Colors, Metrics } from '../Styles/Themes/'
import Icon from 'react-native-vector-icons/FontAwesome'
import { navButtonColor } from '../Styles/NavBarStyle'

type ButtonPropsType = {
  name: string,
  size: number,
  color: string,
  style: any
}

export default {
  navButton (buttonProps: ButtonPropsType, callback: Function) {
    return (
      <TouchableOpacity onPress={callback}>
        <Icon {...defaultButtonProps} {...buttonProps} />
      </TouchableOpacity>
    )
  }
}

const styles = StyleSheet.create({
  button: {
    backgroundColor: Colors.transparent,
    color: navButtonColor,
    justifyContent: 'center'
  }
})

const defaultButtonProps: ButtonPropsType = {
  name: 'heart',
  size: Metrics.icons.medium,
  color: Colors.snow,
  style: styles.button
}
