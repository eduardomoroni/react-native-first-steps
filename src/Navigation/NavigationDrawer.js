// @flow

import React, { PropTypes, Component } from 'react'
import Drawer from 'react-native-drawer'
import { DefaultRenderer, Actions as NavigationActions } from 'react-native-router-flux'
import Menu from './NavigationMenu'
import { connect } from 'react-redux'
import { drawerStyle } from '../Styles/NavBarStyle'

export const menuDrawerProps = {
  key: 'drawer',
  open: false,
  side: 'left',
  type: 'displace',
  content: <Menu />
}

class NavigationDrawer extends Component {
  render () {
    const state = this.props.navigationState
    const defaultProps = {
      ref: 'navigation',
      open: state.open,
      onOpen: () => NavigationActions.refresh({key: state.key, open: true, side: state.side}),
      onClose: () => NavigationActions.refresh(menuDrawerProps),
      content: state.content,
      styles: drawerStyle,
      tapToClose: true,
      openDrawerOffset: 0.25,
      panCloseMask: 0.25,
      negotiatePan: true,
      type: 'overlay',
      side: state.side, // If we set dynamically to right we're getting error
      tweenHandler: tweenHandler
    } // I need to find a better way to override this props in execution time

    return (
      <Drawer {...defaultProps}>
        <DefaultRenderer navigationState={state.children[0]} onNavigate={this.props.onNavigate} />
      </Drawer>
    )
  }
}

const tweenHandler = (ratio) => {
  return {
    main: {
      opacity: Math.max(0.54, 1 - ratio) }
  }
}

NavigationDrawer.propTypes = {
  navigationState: PropTypes.object,
  onNavigate: PropTypes.func
}

export default connect()(NavigationDrawer)
