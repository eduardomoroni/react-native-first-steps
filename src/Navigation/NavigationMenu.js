// @flow

import React, { Component } from 'react'
import PropTypes from 'prop-types'
import { ScrollView, Image, BackAndroid } from 'react-native'
import { connect } from 'react-redux'
import { Images } from '../Styles/Themes'
import { userLoggout } from '../Redux/Actions'
import { DrawerButton } from '../Containers/Components'
import { Actions as NavigationActions } from 'react-native-router-flux'
import I18n from 'react-native-i18n'
import type { Dispatch } from 'redux'

class NavigationMenu extends Component {
  componentDidMount () {
    BackAndroid.addEventListener('hardwareBackPress', () => {
      if (this.context.drawer.props.open) {
        this.toggleDrawer()
        return true
      }
      return false
    })
  }

  toggleDrawer () {
    this.context.drawer.toggle()
  }

  handlePress = (route: string) => {
    this.toggleDrawer()
    NavigationActions[route]()
  }

  renderLoginOrLogout = () => {
    if (this.props.user === null) {
      return (
        <DrawerButton text={I18n.t('login')} onPress={() => this.handlePress('loginScreen')} />
      )
    } else {
      return (
        <DrawerButton text={I18n.t('logout')} onPress={() => this.props.onLogout()} />
      )
    }
  }

  render () {
    return (
      <ScrollView style={styles.container}>
        <Image source={Images.logo} style={styles.logo} />
        <DrawerButton text={I18n.t('cardSearch')} onPress={() => this.handlePress('cardSearchForm')} />
        <DrawerButton text={I18n.t('wishList')} onPress={() => this.handlePress('wishList')} />
        <DrawerButton text={I18n.t('settings')} onPress={() => this.handlePress('settings')} />
        {this.renderLoginOrLogout()}
      </ScrollView>
    )
  }
}

NavigationMenu.contextTypes = {
  drawer: PropTypes.object
}

NavigationMenu.propTypes = {
  user: PropTypes.object,
  onLogout: PropTypes.func
}

const mapStateToProps = (state) => {
  const { token } = state.auth

  return {
    user: token
  }
}

const mapDispatchToProps = (dispatch: Dispatch<*>) => {
  return {
    onLogout: () => dispatch(userLoggout())
  }
}

const styles = {
  container: {
    flex: 1,
    padding: 20
  },
  logo: {
    alignSelf: 'center'
  }
}

export default connect(mapStateToProps, mapDispatchToProps)(NavigationMenu)
