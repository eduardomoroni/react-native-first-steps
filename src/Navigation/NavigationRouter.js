// @flow

import React from 'react'
import { Scene, Router } from 'react-native-router-flux'
import Settings from '../Containers/Settings'
import LoginScreen from '../Containers/LoginScreen'
import MainScreen from '../Containers/MainScreen'
import Styles from '../Styles/NavigationRouterStyle'
import Drawer from './NavigationDrawer'
import I18n from 'react-native-i18n'

const NavigationRouter = () => {
  return (
    <Router>
      <Scene key='drawer' component={Drawer} open={false} >
        <Scene key='drawerChildrenWrapper' navigationBarStyle={Styles.navBar} titleStyle={Styles.title} leftButtonIconStyle={Styles.leftButton} rightButtonTextStyle={Styles.rightButton} >
          <Scene initial key='MainScreen' component={MainScreen} title={I18n.t('welcome')} />
          <Scene key='settings' component={Settings} title={I18n.t('settings')} />
          <Scene key='loginScreen' component={LoginScreen} title={I18n.t('login_for_title')} hideNavBar />
        </Scene>
      </Scene>
    </Router>
  )
}

export default NavigationRouter
