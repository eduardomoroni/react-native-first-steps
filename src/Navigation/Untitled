{
    "name": "two",
    "version": "1.0.0",
    "main": "index.js",
    "license": "MIT",
    "scripts": {
        "echo": "echo \"ONE\" && exit 0"
    }
}
