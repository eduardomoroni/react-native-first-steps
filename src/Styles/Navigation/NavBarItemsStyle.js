import { StyleSheet } from 'react-native'
import { Colors, Metrics } from '../themes'
import { navButtonColor } from '../NavBarStyle'

export default StyleSheet.create({
  button: {
    backgroundColor: Colors.transparent,
    color: navButtonColor,
    justifyContent: 'center',
    marginLeft: Metrics.smallMargin
  }
})
