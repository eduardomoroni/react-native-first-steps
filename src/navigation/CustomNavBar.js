import React from 'react'
import { View, StyleSheet } from 'react-native'
import { builtInBarStyle, customBarStyle } from '../styles/NavBarStyle'

export class CustomNavBar extends React.Component {
  render () {
    const { leftRender, middleRender, rightRender } = this.props

    return (
      <View style={StyleSheet.flatten([customBarStyle.container, builtInBarStyle.navigationBarStyle])}>
        <View style={StyleSheet.flatten([customBarStyle.leftButton, customBarStyle.buttons])}>
          {leftRender}
        </View>
        <View>
          {middleRender}
        </View>
        <View style={StyleSheet.flatten([customBarStyle.rightButton, customBarStyle.buttons])}>
          {rightRender}
        </View>
      </View>
    )
  }
}

export default CustomNavBar
