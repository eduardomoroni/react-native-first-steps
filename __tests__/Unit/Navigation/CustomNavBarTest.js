import { Text } from 'react-native'
import React from 'react'
import { shallow } from 'enzyme'
import { CustomNavBar } from '../../../src/Navigation/CustomNavBar'

const placeholderComponent = (text) => {
  return (
    <Text>{text}</Text>
  )
}

const props = {
  leftRender: placeholderComponent('LEFT'),
  middleRender: placeholderComponent('MIDDLE'),
  rightRender: placeholderComponent('RIGHT')
}

describe('<CustomNavBar />', () => {
  it('Should Render CustomNavBar', () => {
    const wrapper = shallow(<CustomNavBar {...props} />)
    expect(wrapper).toMatchSnapshot()
  })
})

