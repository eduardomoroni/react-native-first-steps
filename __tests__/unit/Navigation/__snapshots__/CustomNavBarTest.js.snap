// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<CustomNavBar /> Should Render CustomNavBar 1`] = `
<View
  style={
    Object {
      "alignItems": "center",
      "backgroundColor": "#3b5998",
      "flexDirection": "row",
      "height": 64,
      "justifyContent": "space-between",
      "left": 0,
      "paddingHorizontal": 10,
      "paddingTop": 5,
      "position": "absolute",
      "right": 0,
      "top": 0,
    }
  }
>
  <View
    style={
      Object {
        "alignItems": "center",
        "flex": 1,
        "flexDirection": "row",
        "justifyContent": "flex-start",
      }
    }
  >
    <Text
      accessible={true}
      allowFontScaling={true}
      ellipsizeMode="tail"
    >
      LEFT
    </Text>
  </View>
  <View>
    <Text
      accessible={true}
      allowFontScaling={true}
      ellipsizeMode="tail"
    >
      MIDDLE
    </Text>
  </View>
  <View
    style={
      Object {
        "alignItems": "center",
        "flex": 1,
        "flexDirection": "row",
        "justifyContent": "flex-end",
      }
    }
  >
    <Text
      accessible={true}
      allowFontScaling={true}
      ellipsizeMode="tail"
    >
      RIGHT
    </Text>
  </View>
</View>
`;
